### ISSUE TITLE
wrap_layout fails on basic site using Ruby 2.5 on Alpine 3.7

### ISSUE THREAD

**Issue by mojavelinux** [middleman/middleman#2150]
## Expected behavior and actual behavior

Middleman fails to build a site using Ruby 2.5 on Alpine 3.7 when the wrap_layout function is used and the layout is a Slim template.

## Steps to reproduce the problem (from a clean middleman installation)

1] Create a new middleman site:

```
$ mkdir site
  cd site
  middleman init
```

2] Add the slim gem to the Gemfile

```
gem 'slim'
```

3] Remove source/layouts/layout.erb

```
rm source/layouts/layout.erb
```

4] Add source/layouts/layout.slim

```
= wrap_layout :default do
  = yield
```

5] Add source/layouts/default.slim

```
doctype 5
html lang='en'
  head
    meta charset='utf-8'
    meta name='viewport' content='width=device-width, initial-scale=1.0'
    title Page Title
  body
    = yield
    = partial '_partials/footer'
```

6] Add source/_partials/_footer.slim

```
= partial '_partials/footer-content'
```

7] Add source/_partials/_footer-content.slim

```
.a : .b : .c : .d : .e : .f : .g : a href='https://example.org/path/to/page/' content
```

8] Build the site

```
bundle exec middleman build
```

9] Observe error

```
Traceback (most recent call last):
	282: from /usr/local/bin/bundle:23:in `<main>'
	281: from /usr/local/bin/bundle:23:in `load'
	280: from /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/exe/bundle:22:in `<top (required)>'
	279: from /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
	278: from /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/exe/bundle:30:in `block in <top (required)>'
	277: from /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/cli.rb:18:in `start'
	276: from /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
	275: from /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.16.1/lib/bundler/cli.rb:27:in `dispatch'
	 ... 270 levels...
	  4: from /site/.bundle/gems/ruby/2.5.0/gems/temple-0.8.0/lib/temple/mixins/dispatcher.rb:7:in `block in on_multi'
	  3: from /site/.bundle/gems/ruby/2.5.0/gems/temple-0.8.0/lib/temple/mixins/dispatcher.rb:49:in `compile'
	  2: from /site/.bundle/gems/ruby/2.5.0/gems/temple-0.8.0/lib/temple/mixins/dispatcher.rb:82:in `dispatcher'
	  1: from /site/.bundle/gems/ruby/2.5.0/gems/temple-0.8.0/lib/temple/mixins/dispatcher.rb:7:in `on_multi'
/site/.bundle/gems/ruby/2.5.0/gems/temple-0.8.0/lib/temple/mixins/dispatcher.rb:7:in `each': stack level too deep (SystemStackError)
bundler: failed to load command: middleman (/site/.bundle/gems/ruby/2.5.0/bin/middleman)
Parallel::DeadWorker: Parallel::DeadWorker
  /site/.bundle/gems/ruby/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:74:in `rescue in work'
  /site/.bundle/gems/ruby/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:71:in `work'
  /site/.bundle/gems/ruby/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:382:in `block (4 levels) in work_in_processes'
  /site/.bundle/gems/ruby/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:495:in `with_instrumentation'
  /site/.bundle/gems/ruby/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:381:in `block (3 levels) in work_in_processes'
  /site/.bundle/gems/ruby/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:369:in `loop'
  /site/.bundle/gems/ruby/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:369:in `block (2 levels) in work_in_processes'
  /site/.bundle/gems/ruby/2.5.0/gems/parallel-1.12.1/lib/parallel.rb:206:in `block (2 levels) in in_threads'
```

## Additional information

- Ruby version: 2.5.0
- Middleman version: 4.2.1
- OS version: Alpine Linux 3.7

The same exact scenario works on Ruby 2.4.

**Comment by mojavelinux** [comment]
It appears that there's a maximum level of nesting the Slim template will support before it goes haywire.

**Comment by mojavelinux** [comment]
This may be an issue w/ Slim or Temple that Middleman is inheriting. The same thing happens if I create a file with ~29 levels of inline nesting:

```
div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div : div
```

Then run that file using `slimrb`.

There's no limit to the amount of nesting I can use on Fedora (I even tried 1000 levels).

**Comment by mojavelinux** [comment]
Here's how I ran Alpine to test:

```
docker run --rm -it ruby:2.5-alpine3.7 ash
```

**Comment by hosh** [comment]
I have been running into this issue elsewhere too (on a Rails app) -- even simpler systems run out of stack. I have tried running a single example of an rspec suite using 100MB of stack and it still fails (but fails in a different area). There's probably a big leak somewhere on the Ruby side of things

I have found the following:

* https://github.com/koshigoe/ruby-2.5.0-stack-level-too-deep
* https://stackoverflow.com/a/48509659

The Stackoverflow answer talks about compiling Ruby against OpenSSL rather than LibreSSL. I will be trying that (not too hard since I'm using a custom Dockerfile based off of the official Ruby Dockerfile). If it works, kinda sucks to be still using OpenSSL, but our team would rather have 2.5 running.
